pandas/core/case_when.py

```diff
@@ -0,0 +1,218 @@
+        is a Series, they must have the same index.
+        If there are multiple replacement options,
+        and they are Series, they must have the same index.
+
+    default : scalar, array-like, default None
+        If provided, it is the replacement value to use
+        if all conditions evaluate to False.
+        If not specified, entries will be filled with the
+        corresponding NULL value.
+
+        .. versionadded:: 2.2.0
+
+    Returns
+    -------
+    Series
+
+    See Also
+    --------
+    Series.mask : Replace values where the condition is True.
+
+    Examples
+    --------
+    >>> df = pd.DataFrame({
+    ...     "a": [0,0,1,2],
+    ...     "b": [0,3,4,5],
+    ...     "c": [6,7,8,9]
+    ... })
+    >>> df
+       a  b  c
+    0  0  0  6
+    1  0  3  7
+    2  1  4  8
+    3  2  5  9
+
+    >>> pd.case_when((df.a.gt(0), df.a),   # condition, replacement
+    ...              (df.b.gt(0), df.b),   # condition, replacement
+    ...              default=df.c)        # optional
+    0    6
+    1    3
+    2    1
+    3    2
+    Name: c, dtype: int64
+    """
+    from pandas import Series
+    from pandas._testing.asserters import assert_index_equal
+
+    args = validate_case_when(args=args)
+
+    conditions, replacements = zip(*args)
+    common_dtypes = [infer_dtype_from(replacement)[0] for replacement in replacements]
+
+    if default is not lib.no_default:
+        arg_dtype, _ = infer_dtype_from(default)
+        common_dtypes.append(arg_dtype)
+    else:
+        default = None
+    if len(set(common_dtypes)) > 1:
+        common_dtypes = find_common_type(common_dtypes)
+        updated_replacements = []
+        for condition, replacement in zip(conditions, replacements):
+            if is_scalar(replacement):
+                replacement = construct_1d_arraylike_from_scalar(
+                    value=replacement, length=len(condition), dtype=common_dtypes
+                )
+            elif isinstance(replacement, ABCSeries):
+                replacement = replacement.astype(common_dtypes)
+            else:
+                replacement = pd_array(replacement, dtype=common_dtypes)
+            updated_replacements.append(replacement)
+        replacements = updated_replacements
+        if (default is not None) and isinstance(default, ABCSeries):
+            default = default.astype(common_dtypes)
+    else:
+        common_dtypes = common_dtypes[0]
+    if not isinstance(default, ABCSeries):
+        cond_indices = [cond for cond in conditions if isinstance(cond, ABCSeries)]
+        replacement_indices = [
+            replacement
+            for replacement in replacements
+            if isinstance(replacement, ABCSeries)
+        ]
+        cond_length = None
+        if replacement_indices:
+            for left, right in zip(replacement_indices, replacement_indices[1:]):
+                try:
+                    assert_index_equal(left.index, right.index, check_order=False)
```

Comment:
From https://github.com/pandas-dev/pandas/pull/55390#discussion_r1391003846

> assert_index_equal ignores order with the `check_order` option, whereas `index.equals` does not ignore order.

Why do we want to ignore order here? If there are multiple orders, which one do we use?